gio/src/subclass/vfs.rs

```diff
@@ -0,0 +1,397 @@
+    }
+}
+
+unsafe extern "C" fn is_active<T: VfsImpl>(vfs: *mut ffi::GVfs) -> glib::ffi::gboolean {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let res = imp.is_active();
+
+    res.into_glib()
+}
+
+unsafe extern "C" fn get_file_for_path<T: VfsImpl>(
+    vfs: *mut ffi::GVfs,
+    path: *const c_char,
+) -> *mut ffi::GFile {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let file = imp.get_file_for_path(&PathBuf::from_glib_none(path));
+
+    file.into_glib_ptr()
+}
+
+unsafe extern "C" fn get_file_for_uri<T: VfsImpl>(
+    vfs: *mut ffi::GVfs,
+    uri: *const c_char,
+) -> *mut ffi::GFile {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let file = imp.get_file_for_uri(&GString::from_glib_borrow(uri));
+
+    file.into_glib_ptr()
+}
+
+unsafe extern "C" fn get_supported_uri_schemes<T: VfsImpl>(
+    vfs: *mut ffi::GVfs,
+) -> *const *const c_char {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let supported_uri_schemes = imp.get_supported_uri_schemes();
+
+    supported_uri_schemes.as_ptr()
+}
+
+unsafe extern "C" fn parse_name<T: VfsImpl>(
+    vfs: *mut ffi::GVfs,
+    parse_name: *const c_char,
+) -> *mut ffi::GFile {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let file = imp.parse_name(&GString::from_glib_borrow(parse_name));
+
+    file.into_glib_ptr()
+}
+
+#[cfg(test)]
+mod tests {
+    // The following tests rely on a custom type `MyLocalVfs` that extends the existing GIO type `GLocalVfs`.
+    // For each virtual method defined in class `gio::ffi::GVfsClass`, a test checks that `MyLocalVfs` and `GLocalVfs` return the same results.
+    // Note that a `MyLocalVfs` instance is built explicitly by calling `glib::Object::builder` whereas a a `GLocalVfs` instance is created by calling `gio::auto::Vfs::local`.
+
+    use super::*;
+    use crate::prelude::*;
+
+    // Binding of existing GIO type GLocalVfs.
+    mod ffi {
+        use crate::ffi;
+
+        #[derive(Copy, Clone)]
+        #[repr(C)]
+        pub struct GLocalVfs {
+            pub parent_instance: ffi::GVfs,
+        }
+
+        #[derive(Copy, Clone)]
+        #[repr(C)]
+        pub struct GLocalVfsClass {
+            pub parent_class: ffi::GVfsClass,
+        }
+    }
+
+    glib::wrapper! {
+        #[doc(alias = "GLocalVfs")]
+        pub struct LocalVfs(Object<ffi::GLocalVfs, ffi::GLocalVfsClass>) @extends Vfs;
+
+        match fn {
+            type_ => || {
+                use std::sync::Once;
+                static ONCE: Once = Once::new();
+
+                // ensure type is initialized by calling `gio::auto::File::for_path` to create a `GLocalFile` instance.
+                ONCE.call_once(|| unsafe {
+                    let _ = File::for_path("path");
+                });
+                glib::gobject_ffi::g_type_from_name("GLocalVfs".to_glib_none().0)
```

Comment:
Not that it really matters because this is a test, but why not call `g_type_from_name()` also inside the `Once` (and make it a `LazyLock`)?